ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/ContractsTable.scala

```diff
@@ -0,0 +1,140 @@
+    // Otherwise, add a delete. This prevents the insertion of transient contracts.
+    def delete(contractId: ContractId, deletion: => Vector[NamedParameter]): AccumulatingBatches =
+      if (insertions.contains(contractId))
+        copy(insertions = insertions - contractId)
+      else
+        copy(deletions = deletions.updated(contractId, deletion))
+
+    private def prepareNonEmpty(
+        query: String,
+        contractIdToParameters: Map[ContractId, Vector[NamedParameter]],
+    ): Option[(Set[ContractId], BatchSql)] = {
+      if (contractIdToParameters.nonEmpty) {
+        val contractIds = contractIdToParameters.keySet
+        val parameters = contractIdToParameters.valuesIterator.toSeq
+        val batch = BatchSql(query, parameters.head, parameters.tail: _*)
+        Some(contractIds -> batch)
+      } else {
+        None
+      }
+    }
+
+    def prepare: PreparedBatches =
+      PreparedBatches(
+        insertions = prepareNonEmpty(insertContractQuery, insertions),
+        deletions = prepareNonEmpty(deleteContractQuery, deletions),
+      )
+
+  }
+
+  def prepareBatchInsert(
+      ledgerEffectiveTime: Instant,
+      transaction: Transaction,
+      divulgedContracts: Iterable[(ContractId, Contract)],
+  ): PreparedBatches = {
+
+    // Add the locally created contracts, ensuring that _transient_
+    // contracts are not inserted in the first place
+    val locallyCreatedContracts =
+      transaction
+        .fold(AccumulatingBatches(Map.empty, Map.empty)) {
+          case (batches, (_, node: Create)) =>
+            batches.insert(
+              contractId = node.coid,
+              insertion = insertContractQuery(
+                contractId = node.coid,
+                templateId = node.coinst.template,
+                createArgument = node.coinst.arg,
+                createLedgerEffectiveTime = Some(ledgerEffectiveTime),
+                stakeholders = node.stakeholders,
+                key = node.key.map(k => Key.assertBuild(node.coinst.template, k.key.value)),
+              )
+            )
+          case (batches, (_, node: Exercise)) if node.consuming =>
+            batches.delete(
+              contractId = node.targetCoid,
+              deletion = deleteContract(node.targetCoid),
+            )
+          case (batches, _) =>
+            batches // ignore any event which is neither a create nor a consuming exercise
+        }
+
+    // Divulged contracts are inserted _after_ locally created contracts to make sure they are
+    // not skipped if consumed in this transaction due to the logic that prevents the insertion
+    // of transient contracts (a divulged contract _must_ be inserted, regardless of whether it's
+    // consumed or not).
+    val divulgedContractsInsertions =
+      divulgedContracts.iterator.collect {
+        case (contractId, contract) if !locallyCreatedContracts.insertions.contains(contractId) =>
+          contractId -> insertContractQuery(
```

Comment:
The query doesn't guard against inserting a divulged contract that has been divulged in previous transactions, i.e. I think we need a `merge` rather than an `insert`.

Reply:
Right, I'll need to work on this a bit, since PostgreSQL and H2 require us to use different idioms to express this. Thanks!

Reply:
Stumbled on this while running the data migration tests.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5541/commits/df4daa6be42762ae7ecf498ef08daa34b386ff55